pkgs/by-name/in/install-grub-ng/src/grub.rs

```diff
@@ -0,0 +1,325 @@
+            };
+            let Some(mut super_options) = fields.next().map(|s| s.split(',')) else {
+                bail!("Super options not found in mountinfo entry: {line}")
+            };
+
+            // Skip the bind-mount on /nix/store.
+            if mount_point == "/nix/store" && super_options.any(|s| s == "rw") {
+                continue;
+            }
+
+            // Skip mountpoint generated by systemd-efi-boot-generator?
+            if fs_type == "autofs" {
+                continue;
+            }
+
+            // Ensure this matches the intended directory
+            if !dir.starts_with(mount_point) {
+                continue;
+            }
+
+            // Is this better than our current match?
+            if mount_point.len() > best.mount.as_os_str().len()
+			  // `is_dir` performs a stat, which can hang forever on network file systems, so
+			  // we only make this call last, when it's likely that this is the mount point.
+				&& Path::new(mount_point).is_dir()
+            {
+                best = Fs {
+                    device: PathBuf::from(device),
+                    fs_type: fs_type.to_owned(),
+                    mount: PathBuf::from(mount_point),
+                }
+            }
+        }
+
+        Ok(best)
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum FsIdentifier {
+    Uuid,
+    Label,
+    Provided,
+}
+impl FsIdentifier {
+    fn to_flag(self) -> &'static str {
+        match self {
+            Self::Uuid => "--fs-uuid",
+            Self::Label => "--label",
+            _ => unreachable!(),
+        }
+    }
+
+    fn to_search(self, fs: &Fs) -> Result<String> {
+        match self {
+            Self::Uuid => self.query_blkid(fs, "UUID"),
+            Self::Label => self.query_blkid(fs, "LABEL"),
+            Self::Provided => Ok(Self::provided_search(&fs.device).unwrap_or_default()),
+        }
+    }
+
+    fn provided_search(device: &Path) -> Option<String> {
+        // If the provided dev is identifying the partition using a label or uuid,
+        // we should get the label / uuid and do a proper search
+        let mut components = device.components();
+        let Some(Component::RootDir) = components.next() else {
+            return None;
+        };
+
+        if components.next() != Some(Component::Normal(OsStr::new("dev"))) {
+            return None;
+        }
+        if components.next() != Some(Component::Normal(OsStr::new("disk"))) {
+            return None;
+        }
+
+        let Some(Component::Normal(by)) = components.next() else {
+            return None;
+        };
+
+        let ty = match by.to_str() {
+            Some("by-label") => Self::Label,
+            Some("by-uuid") => Self::Uuid,
+            _ => return None,
+        };
+
+        Some(format!(
+            "{} {}",
+            ty.to_flag(),
+            components.as_path().display()
+        ))
+    }
```

Comment:
Don't need to mess with components here:
```suggestion
    fn provided_search(device: &Path) -> Option<String> {
        // If the provided dev is identifying the partition using a label or uuid,
        // we should get the label / uuid and ask Grub to do a proper search
        let flag = match device.to_str() {
            Some(s) if s.starts_with("/dev/disk/by-label/") => "--label",
            Some(s) if s.starts_with("/dev/disk/by-uuid/") => "--fs-uuid",
            _ => return None,
        };
    
        Some(format!(
            "{}={}",
            flag,
            device.file_name()?.to_str().expect("path basename must be valid UTF-8")
        ))
    }
```